verifiable/onchain/parser.go

```diff
@@ -0,0 +1,473 @@
+) (*verifiable.CredentialStatus, error) {
+	switch verifiable.CredentialStatusType(credentialStatus.Type) {
+	case verifiable.Iden3OnchainSparseMerkleTreeProof2023:
+		nonce, err := strconv.ParseUint(credentialStatus.RevocationNonce, 10, 64)
+		if err != nil {
+			return nil,
+				fmt.Errorf("invalid revocationNonce '%s': %v", credentialStatus.RevocationNonce, err)
+		}
+
+		chainID, err := core.ChainIDfromDID(*issuerDID)
+		if err != nil {
+			return nil,
+				fmt.Errorf("failed to extract chainID from DID: %v", err)
+		}
+		issuerID, err := core.IDFromDID(*issuerDID)
+		if err != nil {
+			return nil,
+				fmt.Errorf("failed to extract ID from DID: %v", err)
+		}
+		contractAddress, err := core.EthAddressFromID(issuerID)
+		if err != nil {
+			return nil,
+				fmt.Errorf("failed to extract contract address from ID: %v", err)
+		}
+		contractID := fmt.Sprintf("%d:0x%s", chainID, hex.EncodeToString(contractAddress[:]))
+
+		return &verifiable.CredentialStatus{
+			ID: fmt.Sprintf(
+				"%s%s?revocationNonce=%d&contractAddress=%s",
+				issuerDID.String(),
+				credentialStatus.ID,
+				nonce,
+				contractID,
+			),
+			Type:            verifiable.Iden3OnchainSparseMerkleTreeProof2023,
+			RevocationNonce: nonce,
+		}, nil
+	default:
+		return nil,
+			fmt.Errorf("unsupported CredentialStatus type: %s", credentialStatus.Type)
+	}
+
+}
+
+func (p *Parser) convertCredentialSubject(
+	credentialContext []string,
+	fields []CredentialSubjectField,
+) (map[string]any, error) {
+	var (
+		idIsSet   bool
+		typeIsSet bool
+
+		credentialSubject = make(map[string]any)
+	)
+
+	contextbytes, err := json.Marshal(map[string][]string{
+		"@context": credentialContext,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal context: %v", err)
+	}
+
+	// process required fields in credential subject
+	for _, f := range fields {
+		switch f.Key {
+		case "id":
+			idIsSet = true
+			ownderID, ok := big.NewInt(0).SetString(f.Value, 10)
+			if !ok {
+				return nil,
+					fmt.Errorf("failed to extract ownerID '%s' from CredentialSubject", f.Value)
+			}
+			ownerDID, err := bigIntToDID(ownderID)
+			if err != nil {
+				return nil,
+					fmt.Errorf("failed to convert ownerID '%s' to DID: %v", f.Value, err)
+			}
+			credentialSubject["id"] = ownerDID.String()
+		case "type":
+			typeIsSet = true
+			hexBytes, err := hex.DecodeString(hexWithoutPrefix(f.RawValue))
+			if err != nil {
+				return nil,
+					fmt.Errorf("failed to decode hex '%s' from CredentialSubject: %v", f.RawValue, err)
+			}
+			credentialSubject["type"] = string(hexBytes)
+		}
+	}
+
+	if !idIsSet || !typeIsSet {
+		return nil, fmt.Errorf("CredentialSubject does not have required fields: id, type")
+	}
+
+	for _, f := range fields {
+		// skip already processed fields
+		if f.Key == "id" || f.Key == "type" {
+			continue
+		}
+
+		datatype, err := p.marklizeInstance.TypeFromContext(
+			contextbytes,
+			fmt.Sprintf("%s.%s", credentialSubject["type"], f.Key),
```

Comment:
you should use the internal function to determine the type of credential. We can't rely on type inside credentialSubject